agent/consul/state/catalog.go

```diff
@@ -639,8 +643,11 @@ func (s *Store) ensureServiceTxn(tx *memdb.Txn, idx uint64, node string, svc *st
 	if err := tx.Insert("services", entry); err != nil {
 		return fmt.Errorf("failed inserting service: %s", err)
 	}
-	if err := tx.Insert("index", &IndexEntry{"services", idx}); err != nil {
-		return fmt.Errorf("failed updating index: %s", err)
+	if !hasSameTags {
```

Comment:
we can't assume that tags are the only thing that can change. The `catalog/services` end point is also used for [external service registration](https://www.consul.io/docs/guides/external.html) where the node could be different. In the future we may add service metadata and then we have to remember to come here and change the diffing logic. 

What's also not clear to me is why we even need this logic. There's a PR from back in January that addressed the issue of excessive churn by introducing logic to compare even before `ensureServiceTxn` is called https://github.com/hashicorp/consul/commit/f2d9da270d3fb376ce0ada0a1b615a704f7e6c6b . It uses logic defined [here](https://github.com/hashicorp/consul/blob/c2a59f1e6c377e1d99c69809236f1de22e996219/agent/structs/structs.go#L435) to do the comparision. 

The other part of this PR around using `maxIndexForService` as the modify index in the reply looks reasonable, but this conditional !changedTags is too risky to merge

Reply:
@preetapan the point is about X-Consul-Index when using watches as outlined here: https://github.com/hashicorp/consul/issues/3890#issuecomment-366227061 basically, when doing a watch on /catalog/services for instances, if your X-Consul-Index is a bit older (let's say the current raft index of services -1), then it immediately returns the value. (since many changes are changing the "services" index (such as HealthChecks, Services...)

# Why all of this is a concern ?

On a large cluster, we have around 4-5 modifications per second on the "services" index. If the latency between the client is a bit high OR if the client is not performing the watch immediately (for instance, because the implementation is using this result to create other watches), then all the data is downloaded once again and this thing never stops. This is less a problem than for `maxIndexForService` as "services" index if used by far less endpoints (I mean less not in terms of REST endpoints, but in terms of clients watching all those endpoints for changes), but fewer clients are looking at all the services of a node / catalog, but it is still a problem.

I agree with you about tags optimization (and that is why I did not put this optimization when filtering using NodeMeta as well) risk, so I can propose the following:

Add a new index called "services-tags", use it on catalog and similar endpoints, apply the same logic, but with a clear comment in the code.

What do you think ?

Best Regards

Reply:
@pierresouchay we talked about this internally. We think that its best to add a new index, "service-catalog" for this that's updated when the catalog changes according to the logic mentioned above, and with a clear comment documenting why. You will also need to handle deletions similar to what you did in PR #3899 - if the last remaining instance of a service is deregistered it should affect the index value for the `service-catalog` index so watches of the /catalog/services endpoint can see it.

Reply:
@preetapan Ok ! Thank you, I'll do it asap

Reply:
@preetapan DONE